### week2 ~ APIs !

<b>that is, _the-web-as-your-filesystem_ ...</b>  &nbsp;&nbsp; (hw2pr1.ipynb)

[the google doc with hw2's details](https://docs.google.com/document/d/1z4HwpUL1-ImGX3j-6bddyZAfdv2SkRQHRNhF5Cu3Fkk/edit?tab=t.0)
<hr>

In [3]:
#
# computing-styling trick of the day     (or, of the minute...)
#
# The setting for word-wrapping on the output is
#     "notebook.output.wordWrap": true,   (in your settings.json file or from Code ... Settings ...) 

print( list(range(100)) )


[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99]


#### Let's see if you already have the requests library...

In [4]:
#
# see if we have the requests library...
#

import requests

In [5]:
#
# If you _don't_ have the requests library, let's install it!
#

# for me, it worked to uncomment and run this command, here in this cell:
# !pip3 install requests  OR   !pip install requests

# an alternative is to run, in a terminal, the command would be 
#  pip3 install requests  OR    pip install requests      (the ! is needed only if inside Python)

# It's very system-dependent how much you have to "restart" in order to use
# the new library (the notebook, VSCode, the Jupyter extension, etc.)

# Troubles?  Let us know!  We'll work on it with you...

!pip install requests


[notice] A new release of pip is available: 25.0 -> 25.0.1
[notice] To update, run: pip install --upgrade pip


In [6]:
#
# hopefully, this now works! (if so, running will succeed silently)
#

import requests

In [7]:
#
# let's try it on a simple webpage
#

#
# we assign the url and obtain the api-call result into result
#    Note that result will be an object that contains many fields (not a simple string)
# 

url = "https://www.cs.hmc.edu/~dodds/demo.html"
result = requests.get(url)
result    

# if it succeeded, you should see <Response [200]>
# See the list of HTTP reponse codes for the full set!

<Response [200]>

[Here is Wikipedia's list of all HTTP response codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes)
+ 100's: information
+ 200's: success           
+ 300's: redirects
+ 400's + 500's: errors

Perhaps familiar:  404 <br>
Especially fun:  418


In [8]:
#
# when exploring, you'll often obtain an unfamiliar object. 
# Here, we'll ask what type it is 
type(result)

requests.models.Response

How to access the data inside this object, ``result`` ?

One way: [Head over to the online documentation](https://requests.readthedocs.io/en/latest/api/#requests.Response)

In addition, you can "look around" in Python:

In [9]:
# here is one of the data members within the result
# it "remembers" (keeps track of) the url requested:
result.url

'https://www.cs.hmc.edu/~dodds/demo.html'

In [10]:
# We can print all of the data members in an object with dir
# Since dir returns a list, we will grab that list and loop over it:
all_fields = dir(result)

for field in all_fields:
    if "_" not in field: 
        print(field)

close
connection
content
cookies
elapsed
encoding
headers
history
json
links
next
ok
raw
reason
request
text
url


In [11]:
#
# Let's try printing a few of those fields (data members): 
print(f"result.url         is {result.url}")  # the original url
print(f"result.raw         is {result.raw}")  # another object!
print(f"result.encoding    is {result.encoding}")  # utf-8 is very common
print(f"result.status_code is {result.status_code}")  # 200 is success!

result.url         is https://www.cs.hmc.edu/~dodds/demo.html
result.raw         is <urllib3.response.HTTPResponse object at 0x106a0ce80>
result.encoding    is ISO-8859-1
result.status_code is 200


In [12]:
# In this case, the result is a text file (HTML) Let's see it!
contents = result.text
print(contents)

<html>
  <head>
    <title>My streamlined website</title>
  </head>
  <body>
    <h1> Welcome! </h1>
    <h2> The best numbers </h2>

    <div id="numberlist">
      <ol>
	<li class="number"> 35 </li>
	<li class="number"> 42 </li>
	<li class="number"> <a href="https://en.wikipedia.org/wiki/Rayo%27s_number">Rayo's number</a> </li>
      </ol>
    </div>

    <img src="./spam.jpg" height="84px">
    <br><br>

    <h2> The <s>only</s> best snacks </h2>

    <div id="snacklist">
      <ul>
	<li class="snack"> Poptarts </li>
	<li class="snack"> Chocolate </li>
	<li class="snack"> Coffee </li>
      </ul>
    </div>

<!--    <a href="./demo_cat.html">Aliens <3 cats!</a>  -->

    <img src="./alien.png" height="101px">

  </body>
</html>






In [13]:
# Yay!  
# This shows that you are able to "scrape" an arbitrary HTML page... 

# Now, we're off to more _structured_ data-gathering...

#### Traversing the world - and web - <i>without a browser</i>.   

<b>Using the ISS APIs</b> 

+ Here, you'll make some calls using `requests` to, first, the International Space Station API 
+ and, then, the US Geological Survey's earthquake API
+ "API" is short for "Application Programming Interface" 
  + Admittedly, this is not a very informative name:
  + The API is the set of services, which are functions and/or urls, provided by some software or site

Let's try it with the International Space Station api at [http://api.open-notify.org/iss-now.json](http://api.open-notify.org/iss-now.json)
+ [This page has documentation on the ISS API](http://open-notify.org/Open-Notify-API/ISS-Location-Now/)

In [14]:
#
# we assign the url and obtain the api-call result into result
#    Note that result will be an object that contains many fields (not a simple string)
# 

import requests

url = "http://api.open-notify.org/iss-now.json"   # this is sometimes called an "endpoint" ...
result = requests.get(url)
result    

# if it succeeds, you should see <Response [200]>

<Response [200]>

In [15]:
#
# Let's try printing those shorter fields from before:
print(f"result.url         is {result.url}")  # the original url
print(f"result.raw         is {result.raw}")  # another object!
print(f"result.encoding    is {result.encoding}")  # utf-8 is very common
print(f"result.status_code is {result.status_code}")  # 200 is success!

result.url         is http://api.open-notify.org/iss-now.json
result.raw         is <urllib3.response.HTTPResponse object at 0x10c0ab9d0>
result.encoding    is utf-8
result.status_code is 200


In [16]:
#
# In this case, we know the result is a JSON file, and we can obtain it that way:
json_contents = result.json()
print(json_contents)

# Remember:  json_contents will be a _dictionary_

{'message': 'success', 'timestamp': 1739429843, 'iss_position': {'longitude': '129.0454', 'latitude': '4.6847'}}


In [17]:
#
# Let's re-remind ourselves how dictionaries work:

json_contents['iss_position']['longitude']       # Challenge:  could we access the other components? What _types_ are they?!!

'129.0454'

In [18]:
#
# In Python, we can use the resulting dictionary... let's see its keys:
print(list(json_contents.keys()))  

# Also, watch out for string vs. numeric types, e.g., for latitude and longitude.
# At heart, _all_ web data are strings... .

# These experiments will be helpful for problem 1, below :)

['message', 'timestamp', 'iss_position']


<font color="Coral"><b>Task 0a</b></font>  &nbsp;&nbsp; From here, 
+ extract the lat and CS35_Participant_2 of the ISS from ``json_contents``  (dictinoary practice!)
+ copy the ``haversine`` function from the assignment
+ find the lat/CS35_Participant_2 of Claremont (or grab it from the hw page! 😊 )
+ combine these to find out how far, in miles, the ISS is from Claremont  
+ <font size="-2">you could _imagine_ writing a function to do all this (no need to - we'll do it with earthquakes!)</font>

In [19]:
import requests
from math import *

def haversine(lat1, long1, lat2, long2):
    """
    Calculate the great circle distance in kilometers between two points
    on the earth (specified in decimal degrees)
    """
    # convert decimal degrees to radians
    long1, lat1, long2, lat2 = map(radians, [long1, lat1, long2, lat2])


    # haversine formula
    dlong = long2 - long1
    dlat = lat2 - lat1
    trig = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlong/2)**2
    # Radius of earth. Use 3956 for miles. 6371 for km.
    radius = 3956  # we'll use miles!
    return radius * 2 * asin(sqrt(trig))

url = "http://api.open-notify.org/iss-now.json"  
# result = requests.get(url)
# result    
# <Response [200]>!

#ISS latitude & longitude
ISSlat = float(json_contents['iss_position']['latitude'])
ISSlong = float(json_contents['iss_position']['longitude'])

#Claremont latitude & longitude
Clarelat = 34.0967
Clarelong = -117.7198

print(f"The ISS is {haversine(ISSlat, ISSlong, Clarelat, Clarelong)} miles away from Claremont!")


The ISS is 7335.962117538471 miles away from Claremont!


## JSON

####  Let's make a brief JSON visit in Python

The library ``json`` allows us to create and read arbitrary JSON data.

In [20]:
# JSON is a javascript dictionary format -- almost the same as a Python dictionary:
data = { 'key':'value',  'fave':42,  'list':[5,6,7,{'mascot':'Aliiien'}] }
print(data)

# we can write in JSON format to a local file, named small42.json:
import json 

with open("small.json", "w") as f:
    json.dump( data, f )

{'key': 'value', 'fave': 42, 'list': [5, 6, 7, {'mascot': 'Aliiien'}]}


In [21]:
# We can also read from a json file
# The resulting data will be a _dictionary_:

with open("small.json", "r") as f:
    dictionary = json.load( f )

print(f"the {dictionary = }")

the dictionary = {'key': 'value', 'fave': 42, 'list': [5, 6, 7, {'mascot': 'Aliiien'}]}


In [22]:
# let's access this dictionary -- first, the keys:
list(dictionary.keys())   # How do we get 'Aliiien' from newdata?

['key', 'fave', 'list']

In [23]:
# Task: use the dictionary to obtain (a) 'value' , (b) 42 , (c) 'Aliiien'  [tricky!]

# remember that there are two ways to get the value from a key:
# way 1:  dictionary['key']            # errors if 'key' isn't present
# way 2:  dictionary.get('key')        # returns None if 'key' isn't present

dictionary['key']     # value
dictionary.get('fave')
dictionary['list'][3]['mascot']


'Aliiien'

Most of the time this will be done for us by the ``requests`` library. So, we will simply receive the dictionary of data sent.

Then, the trick is to "extract" the data fragments we want. (Sometimes it feels like forensics - or archaeology!) Try excavating items one **layer** at a time...

#### Remember: &nbsp; <i>not</i> every url returns json data...
+ The url [https://www.cs.hmc.edu/~dodds/demo.html](https://www.cs.hmc.edu/~dodds/demo.html) returns a plain-text file with _markup_ text
+ that is to say, with HTML tags, such as `<title>Title</title>` to designate the components of its content
+ HTML stands for _hypertext markup language_   
+ Often anything with tags similar to `<b>be bold!</b>` is called "markup." 

Let's try our 5C homepages: they're HTML, not JSON:

In [24]:
import requests 

# here, we will obtain plain-text results from a request
#url = "https://www.cs.hmc.edu/~dodds/demo.html"  # try it + source
# url = "https://www.scrippscollege.edu/"          # another possible site...
# url = "https://www.pitzer.edu/"                  # another possible site...
url = "https://www.cmc.edu/"                     # and another!
# url = "https://www.cgu.edu/"
result = requests.get(url)        
print(f"result is {result}")        # hopefully it's 200

result is <Response [200]>


In [25]:
# if the request was successful, the Response will be [200]. 
# Then, we can grab the text - or json - from the site:

text = result.text                  # provides the HTML page as a large string...
print(f"len(text) is {len(text)}")  # let's see how large the HTML page is... 

#print("\nThe first 242 characters are\n")
print(text[:])                  # we'll print the first few characters...  

# change this to text[:] to see the whole document...
# Notice that we can run many different analyses without having to re-call/re-scrape the page (this is good!)

len(text) is 362390
<!DOCTYPE html>
<html lang="en" dir="ltr" prefix="og: https://ogp.me/ns#">
  <head>
    <meta charset="utf-8" />
<meta name="description" content="Claremont McKenna College, one of the nation’s top liberal arts colleges, prepares students to make a difference. The College educates students for thoughtful and productive lives and roles of responsible leadership through its strong, real-world mission. Admission to the College is need-blind" />
<meta name="keywords" content="Claremont McKenna College, college, CMC" />
<link rel="shortlink" href="https://www.cmc.edu/" />
<link rel="canonical" href="https://www.cmc.edu/" />
<meta name="google-site-verification" content="google3f3dd1452ae49f8c.html" />
<meta name="Generator" content="Drupal 10 (https://www.drupal.org)" />
<meta name="MobileOptimized" content="width" />
<meta name="HandheldFriendly" content="true" />
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<link rel="icon" href="/themes/cus

#### Declined requests!

Caution: <i>See this week's reading!</i>

#### For now let's focus on API calls providing JSON 
+ We're reading-aligned, as we should be!
+ The open-ended problem (the finale in this notebook) offers the _option_ of scraping raw html -- up to you...

<br>

<b>Let's try another ISS "endpoint" ~ one with all of the <i>people</i> in space.</b>

It's at this url:  [http://api.open-notify.org/astros.json](http://api.open-notify.org/astros.json)


In [26]:
#
# we assign the url and use requests.get to obtain the result into result_astro
#
#    Remember, result_astro will be an object that contains many fields (not a simple string)
# 

import requests

url = "http://api.open-notify.org/astros.json"   # this is sometimes called an "endpoint" ...
result_astro = requests.get(url)
result_astro

# if it succeeded, you should see <Response [200]>

<Response [200]>

In [27]:
# If the request succeeded, we know the result is a JSON file, and we can obtain it that way.
# Let's call our dictionary something more specific:

astronauts = result_astro.json()
print(astronauts)
d = astronauts     # d is shorter to type

# Remember:  d and astronauts will be a _dictionary_

note = """ here's yesterday's result - it _should_ be the same today!

{"people": [{"craft": "ISS", "name": "Oleg Kononenko"}, {"craft": "ISS", "name": "Nikolai Chub"},
{"craft": "ISS", "name": "Tracy Caldwell Dyson"}, {"craft": "ISS", "name": "Matthew Dominick"},
{"craft": "ISS", "name": "Michael Barratt"}, {"craft": "ISS", "name": "Jeanette Epps"},
{"craft": "ISS", "name": "Alexander Grebenkin"}, {"craft": "ISS", "name": "Butch Wilmore"},
{"craft": "ISS", "name": "Sunita Williams"}, {"craft": "Tiangong", "name": "Li Guangsu"},
{"craft": "Tiangong", "name": "Li Cong"}, {"craft": "Tiangong", "name": "Ye Guangfu"}], "number": 12, "message": "success"}
"""
""

{'people': [{'craft': 'ISS', 'name': 'Oleg Kononenko'}, {'craft': 'ISS', 'name': 'Nikolai Chub'}, {'craft': 'ISS', 'name': 'Tracy Caldwell Dyson'}, {'craft': 'ISS', 'name': 'Matthew Dominick'}, {'craft': 'ISS', 'name': 'Michael Barratt'}, {'craft': 'ISS', 'name': 'Jeanette Epps'}, {'craft': 'ISS', 'name': 'Alexander Grebenkin'}, {'craft': 'ISS', 'name': 'Butch Wilmore'}, {'craft': 'ISS', 'name': 'Sunita Williams'}, {'craft': 'Tiangong', 'name': 'Li Guangsu'}, {'craft': 'Tiangong', 'name': 'Li Cong'}, {'craft': 'Tiangong', 'name': 'Ye Guangfu'}], 'number': 12, 'message': 'success'}


''

This is pretty intricate. Let's try unpacking this - _parsing it_ - with an in-class break-out challenge.

In [28]:
#
# Cell to try out parsing d  (astronauts)
#

d['people'][0]['name'][7:9]

'no'

Let's compare with a whole other webservice: **earthquakes** 

<br>
<hr>
<br>

#### Earthquake data

[Here is the USGS Earthquate data API documentation](https://earthquake.usgs.gov/fdsnws/event/1/)

Notice that the "headline" is the URL: &nbsp; This is the <i>domain</i> from which we'll access data.
+ Underneath, there are several different <i>endpoints</i> 
+ We are going to focus on the <tt>count</tt> endpoint and the <tt>query</tt> endpoint
+ along with their parameters
  + the whole list of parameters is linked and available by scrolling down
  + that said, it's easy to miss (well, at least I did! :)

First, let's establish that, for these endpoints, we can make requests - by hand - in our browser!

Try this link: <br><br>  [https://earthquake.usgs.gov/fdsnws/event/1/count?format=geojson&minmagnitude=5.0&starttime=2024-01-01&endtime=2024-02-01](https://earthquake.usgs.gov/fdsnws/event/1/count?format=geojson&minmagnitude=5.0&starttime=2024-01-01&endtime=2024-02-01)

Ok! Let's parse this url. Requests in which there are parameters <i><b>in the url</b></i> are called GET requests:
+ the <b>endpoint</b> is the first part: ``https://earthquake.usgs.gov/fdsnws/event/1/count``
   + Notice that the forward-slashes are very much like our file-system trees from last week!
   + In fact, they usually <i>are</i> file-system trees. They're just on the <i>server</i> side, instead of our "client" side...
+ the <b>parameters</b> follow the question mark: ``format=geojson&minmagnitude=5.0&starttime=2024-01-01&endtime=2024-02-01``
   + There are four parameters here. Parameters are separated by the ampersand <tt>&amp;</tt> character.
   + Each one is in the format <tt>name=value</tt>  Here are the four:
   + ``format=geojson`` specifies the desired format of the returned data. ``geojson`` is JSON with geographic data.
   + ``minmagnitude=5.0`` specifies the minimum magnitude of earthquakes to consider. 5.0 is strong, if not always catastrophic.
   + ``starttime=2024-01-01`` specifies the earlier time-endpoint to consider. (Jan 1, 2024)
   + ``endtime=2024-02-01`` specifies the later time-endpoint to consider. (Feb 1, 2024)

My result was this:  ``{"count":134,"maxAllowed":20000}``
+ Earthquakes are always happening -- and they do get reclassified. So, the numbers can change - even in the past.

Try it, in your browser.

Also, try increasing the ``minmagnitude`` -- you'll see a progression of fewer and fewer quakes. (Fortunately!) 

<hr>

Then, try it using a short Python script:

In [29]:
#
# Let's try the  count  endpoint, with geojson format (json with geographical data)
#

url = "https://earthquake.usgs.gov/fdsnws/event/1/count?format=geojson&minmagnitude=5.0&starttime=2024-01-01&endtime=2024-02-01"

result = requests.get(url)                       # a named input, params, taking the value param_d, above
print(f"result is {result}")                     # hopefully, this is 200
print(f"the full url used was\n {result.url}")   # it's nice to be able to see this

result is <Response [200]>
the full url used was
 https://earthquake.usgs.gov/fdsnws/event/1/count?format=geojson&minmagnitude=5.0&starttime=2024-01-01&endtime=2024-02-01


In [30]:
# If it worked, we should be able to obtain the JSON. Remember, it's a dictionary. Let's use d:

d = result.json()

print(f"{d =}")

d ={'count': 133, 'maxAllowed': 20000}


#### Handling parameters separately...

It's awkward to include all the parameters as part of the url.

It's much more common to create a <i>dictionary</i> of the parameters, and then pass that to ``requests.get``

Here is an example:

In [31]:
#
# Here is the endpoint
#
url = "https://earthquake.usgs.gov/fdsnws/event/1/count"

# Let's use variables for three of the parameters:
min_mag = 5.0               # the minimum magnitude considered a quake (min_mag)
start_time = "2025-01-01"   # this is the year-month-day format of the start
finish_time = "2025-02-01"  # similar for the end

# we assemble a dictionary of our parameters, let's name it param_dictionary
# there are many more parameters available. The problems below ask you to explore them...
param_dictionary = { "format":"geojson",         # this is simply hard-coded to obtain json
                     "starttime":start_time,
                     "endtime":finish_time,
                     "minmagnitude":min_mag,
                     }

# Here, we use requests to make the request. The parameters will be added by this API call:
result = requests.get(url, params=param_dictionary)
print(f"result is {result}")                     # hopefully, this is 200
print(f"the full url used was\n {result.url}")   # this will include the parameters!

result is <Response [200]>
the full url used was
 https://earthquake.usgs.gov/fdsnws/event/1/count?format=geojson&starttime=2025-01-01&endtime=2025-02-01&minmagnitude=5.0


In [32]:
# If it worked, we should be able to see the json results:

d = result.json()
print(f"JSON returned was {d = }")

JSON returned was d = {'count': 150, 'maxAllowed': 20000}


From here, it would be possible to write one or more loops and build an earthquake dataset. For example,
+ it would be possible to loop over the ``minmagnitude`` to get a distribution of different sized quakes (or a histogram)
+ it would be possible to loop over one of the <i>time-endpoints</i>
+ it would be possible to loop over one of the _other parameters_ e.g.,
  + you can specify a circle around a specific ``latitude`` and ``longitude`` with a ``maxradiuskm`` (the radius)
  + Claremont is at ``latitude=34.0967`` and ``longitude=-117.7198`` 

The next two cells have an example of a Claremont-centric quake-count question and answer:

In [33]:
#
# How many quakes of magnitude >= 4.2 have been within 300km of Claremont 
#     + in Jan 2025
#     + in Dec 2025
#
url = "https://earthquake.usgs.gov/fdsnws/event/1/count"

# Let's use variables for three of the parameters:
min_mag = 2.2               # the minimum magnitude considered a quake (min_mag)
start_time = "2025-01-01"   # this is the year-month-day format of the start
finish_time = "2025-02-01"  # similar for the end
# start_time = "2024-01-01"   # similar, but for a year-CS35_Participant_2 span...
# finish_time = "2025-01-01"  # similar for the end
radius_in_km = 300

# we assemble a dictionary of our parameters, let's name it param_dictionary
# there are many more parameters available. The problems below ask you to explore them...
param_dictionary = { "format":"geojson",         # this is simply hard-coded to obtain json
                     "starttime":start_time,
                     "endtime":finish_time,
                     "minmagnitude":min_mag,
                     "latitude":34.0967,
                     "longitude":-117.7198,
                     "maxradiuskm":radius_in_km,
                     }

# Here, we use requests to make the request. The parameters will be added by this API call:
result = requests.get(url, params=param_dictionary)
print(f"result is {result}")                     # hopefully, this is 200
print(f"the full url used was\n {result.url}")   # this will include the parameters!

# We'll extract the final result in another cell:

result is <Response [200]>
the full url used was
 https://earthquake.usgs.gov/fdsnws/event/1/count?format=geojson&starttime=2025-01-01&endtime=2025-02-01&minmagnitude=2.2&latitude=34.0967&longitude=-117.7198&maxradiuskm=300


In [34]:
# Let's finish up here:
quake_count = result.json()
print(f"{quake_count = }")

quake_count = {'count': 70, 'maxAllowed': 20000}


<font color="Coral"><b>Try this, task "0b"</b></font>  &nbsp;&nbsp; 
+ Would you expect more or fewer quakes if a minimum magnitude of 2.2 were used?
+ Try the above cells again for a minimum magnitude of 2.2 -- if that the difference you'd expect?

##### 0b response part 1
+ I expect way more quakes since a minimum of 2.2 encompasses all those above 4.2 anyway, and we get minor quakes all the time.
+ I was right! The number of quakes jumped from 1 to 69, which is a pretty significant difference.

<font color="DodgerBlue"><b>Quake-counting Results:</b></font> 

#### Number of Claremont-centric quakes
  + <u>Overall</u> The API calls to the USGS showed that, within 300km of Claremont, there were
    + 1 quake of magnitude 4.2 or larger within 300km of Claremont in Jan '25
    + more quakes  of magitude 2.2 or larger within 300km of Claremont in Jan '25 <br><br>
  + <u>Reflection</u>: _This is not enough data to establish a trend. (I hope!)_ That said, for the hw, you should include a **loop** over at least 10 values, as well as a text-formatted set of values. (It's ok for the output values to be in the computational cell(s), not the markdown cell. The markdown is really for reflection on results than reporting of results.) <br><br>
  + <u>Opportunities</u>: In fact, there are lots of values over which the USGS API allows to vary. For example,  minmagnitude (or maxmagnitude), radius, location (lat/CS35_Participant_2), months (or other time-measurements) - and others. In addition, there is the chance to look at the details of each quake using the ``query`` endpoint. <br><br>

Looping over API calls to gather data will be one of the hw problems.

<hr>

#### The ``query`` endpoint

Let's see, too, that it's possible to obtain, not only a <i>count</i>, but also a <b>"full report"</b> of all of the earthquakes.

To do so, the only change needed is from the endpoint ``count`` to the endpoint ``query``

First, try it "by hand" -- by opening this url in your browser:

[https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&minmagnitude=6.8&starttime=2024-01-01&endtime=2024-02-01](https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&minmagnitude=6.8&starttime=2024-01-01&endtime=2024-02-01)

Take a look -- there is a lot more data!  

Next, let's try it programmatically:

In [143]:
#
# Here is the endpoint
#
url = "https://earthquake.usgs.gov/fdsnws/event/1/query"

# Let's use variables for three of the parameters:
min_mag = 6.5               # the minimum magnitude considered a quake (min_mag)
start_time = "2025-01-01"   # this is the year-month-day format of the start
finish_time = "2025-02-10"  # similar for the end

# we assemble a dictionary of our parameters, let's name it param_dictionary
# there are many more parameters available. The problems below ask you to explore them...
param_dictionary = { "format":"geojson",         # this is simply hard-coded to obtain json
                     "starttime":start_time,
                     "endtime":finish_time,
                     "minmagnitude":min_mag,
                     }

# Here, we use requests to make the request. The parameters will be added by this API call:
result = requests.get(url, params=param_dictionary)
print(f"result is {result}")                     # hopefully, this is 200
print(f"the full url used was\n {result.url}")   # this will include the parameters!

result is <Response [200]>
the full url used was
 https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&starttime=2025-01-01&endtime=2025-02-10&minmagnitude=6.5


In [78]:
# If it worked, we should be able to see the json results:

d = result.json()
print(f"JSON returned was {d = }")

JSON returned was d = {'type': 'FeatureCollection', 'metadata': {'generated': 1739219790000, 'url': 'https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&starttime=2025-01-01&endtime=2025-02-10&minmagnitude=6.5', 'title': 'USGS Earthquakes', 'status': 200, 'api': '1.14.1', 'count': 3}, 'features': [{'type': 'Feature', 'properties': {'mag': 7.6, 'place': '209 km SSW of George Town, Cayman Islands', 'time': 1739056994341, 'updated': 1739198992069, 'tz': None, 'url': 'https://earthquake.usgs.gov/earthquakes/eventpage/us7000pcdl', 'detail': 'https://earthquake.usgs.gov/fdsnws/event/1/query?eventid=us7000pcdl&format=geojson', 'felt': 390, 'cdi': 7.3, 'mmi': 4.901, 'alert': 'green', 'status': 'reviewed', 'tsunami': 1, 'sig': 1173, 'net': 'us', 'code': '7000pcdl', 'ids': ',pt25039000,at00sre0yr,us7000pcdl,usauto7000pcdl,', 'sources': ',pt,at,us,usauto,', 'types': ',dyfi,finite-fault,general-text,ground-failure,impact-link,internal-moment-tensor,internal-origin,losspager,moment-tens

In [79]:
#
# That's hard to read!
# Let's pretty-print it with the json library
#       Also, this version can be pasted into online formatters, e.g., https://jsonformatter.org/

import json 
nice_string = json.dumps(d)   # this outputs a "nicely formatted string" using double quotes
print(nice_string)



{"type": "FeatureCollection", "metadata": {"generated": 1739219790000, "url": "https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&starttime=2025-01-01&endtime=2025-02-10&minmagnitude=6.5", "title": "USGS Earthquakes", "status": 200, "api": "1.14.1", "count": 3}, "features": [{"type": "Feature", "properties": {"mag": 7.6, "place": "209 km SSW of George Town, Cayman Islands", "time": 1739056994341, "updated": 1739198992069, "tz": null, "url": "https://earthquake.usgs.gov/earthquakes/eventpage/us7000pcdl", "detail": "https://earthquake.usgs.gov/fdsnws/event/1/query?eventid=us7000pcdl&format=geojson", "felt": 390, "cdi": 7.3, "mmi": 4.901, "alert": "green", "status": "reviewed", "tsunami": 1, "sig": 1173, "net": "us", "code": "7000pcdl", "ids": ",pt25039000,at00sre0yr,us7000pcdl,usauto7000pcdl,", "sources": ",pt,at,us,usauto,", "types": ",dyfi,finite-fault,general-text,ground-failure,impact-link,internal-moment-tensor,internal-origin,losspager,moment-tensor,origin,phase-data,s

but, we can do better! 

the "dump string" function, json.dumps, can output the formatted version, too...


In [80]:
import json 
nicer_string = json.dumps(d, indent=4)   # We can specify the indentation. 
print(nicer_string)                      # It's another tree structure... !

{
    "type": "FeatureCollection",
    "metadata": {
        "generated": 1739219790000,
        "url": "https://earthquake.usgs.gov/fdsnws/event/1/query?format=geojson&starttime=2025-01-01&endtime=2025-02-10&minmagnitude=6.5",
        "title": "USGS Earthquakes",
        "status": 200,
        "api": "1.14.1",
        "count": 3
    },
    "features": [
        {
            "type": "Feature",
            "properties": {
                "mag": 7.6,
                "place": "209 km SSW of George Town, Cayman Islands",
                "time": 1739056994341,
                "updated": 1739198992069,
                "tz": null,
                "url": "https://earthquake.usgs.gov/earthquakes/eventpage/us7000pcdl",
                "detail": "https://earthquake.usgs.gov/fdsnws/event/1/query?eventid=us7000pcdl&format=geojson",
                "felt": 390,
                "cdi": 7.3,
                "mmi": 4.901,
                "alert": "green",
                "status": "reviewed",
         

<font color="Coral"><b>Try this, also for Task 0b</b></font>  &nbsp;&nbsp; 
+ Look through the response to find where and when the largest eathquake was, in the previous year
+ What was its magnitude?
+ (optional) Do you see a way that - just by making small changes and re-running - you could find the second-biggest quake of '24?  Perhaps try it...  🦔

##### 0b response part 2
+ the largest earthquake of 2025 so far was the earthquake 209 km from George Town, Cayman Islands
+ February 8th (geojson time is given in milliseconds after 1970-01-01, 00:00:00.000, so I just calculated it)
+ its magnitude was 7.6
+ change start and end time to 2024-01-01 and 2025-01-01, run with minimum around 7.5 as there won't be many outputs (knowing earthquake trends)

<br>
<hr>
<br>



#### Launching into hw2's challenges:
+ <font color="Coral"><b>Tasks1-2</b></font> &nbsp;&nbsp; ISS challenges:
  + ``ISS_now()`` which will find and return the ISS's lat/CS35_Participant_2 (as floats)
    + use the earlier examples as a starting point...
  + ``ISS_distance()`` which will return the distance of the ISS from a city of your choice (a constant city, not a variable). It can be Claremont, but doesn't have to be. This will require using the <i>haversine</i> distance for global lat/CS35_Participant_2 coordinates...
+ <font color="Coral"><b>Tasks3-4</b></font> &nbsp;&nbsp; Earthquake challenges:
  + ``Quake_loop()`` which will loop over a parameter of your choice, print a formatted list of quake-count data, and return that list
  + ``Quake_compare(place1, place2)`` which will ask - and answer - a comparative question about "quakiness" for two different places, using the quake data... 
    + Here, the goal is to define which of the two places is "quakiest."
    + Notice that the definition of "quakiest" is up to you...
    + The inputs, place1 and place2, can be lat/CS35_Participant_2 pairs - or strings, which then get looked up...

<br>

#### <font color="Coral"><b>Task5</b></font> &nbsp;&nbsp; Open-ended, two-hop challenge

Then, you'll choose or create an open-ended API task -- or create a variant of your own design from at least two APIs - or "scrapes" - of your choice.
+ The key constraint is to be sure you meaningfully use the  <font color="DodgerBlue">result</font>  from the first API call in order to customize the request of the second API call.
+ Then, with the result of that second API call, interpret the data obtained to make a final "statement" -- which can be anywhere amid serious, silly, surreal -- that combines the two API insights.
+ See below for a far-fetched superbowl-themed idea...




In [93]:
# ISS now
import requests

def ISS_now(url): 
    resultnow = requests.get(url) 
    ISScoord = resultnow.json()  
    # checked that data was successfully retrieved.. <Response [200]>!

    #ISS latitude & longitude
    ISSlat = float(ISScoord['iss_position']['latitude'])         # float allows function to be used below
    ISSlong = float(ISScoord['iss_position']['longitude'])
    
    return ISSlat, ISSlong

url = "http://api.open-notify.org/iss-now.json" #ISS now endpoint

ISSlat, ISSlong = ISS_now(url)

print(f"The current latitude of the ISS is {ISSlat} and the current longitude is {ISSlong}")


The current latitude of the ISS is 6.6298 and the current longitude is -126.6069


In [113]:
# ISS distance
from haversine import haversine
import requests

def ISS_distance(city):
    url = "http://api.open-notify.org/iss-now.json"  # same endpoint as above
    ISSlat, ISSlong = ISS_now(url)                   # extract latitude and longitude
    ISS = [ISSlat, ISSlong]

    # I tried to allow for users to input a city, but I started running into problems with API keys and websites blocking automated API calls :(
    # cityurl = "https://nominatim.openstreetmap.org/search?q="+placeinput+"&format=json"
    # city_req = requests.get(cityurl)
    # city_result = city_req.json() 

    # citylat = city_result['lat']
    # citylong = city_result['lon']

    return haversine(ISS, city)


# placeinput = input("To which city do you want to find the distance to the ISS? ")
# since the above ^^^ didn't work...
# instead I went with London 
londonlat = 51.509865
londonlong = -0.118092
london = [londonlat, londonlong]

distance = ISS_distance(london)

print(f"The distance from London to the ISS is {distance} kilometers")

The distance from London to the ISS is 16232.173997819858 kilometers


#### USGS Challenges

Tasks 3 and 4 use the earthquake API

In [36]:
# Quake_loop()
import requests
import time

def Quake_loop(year, lat, lon):
   """ returns largest quake in a year in any area"""
   
   url = "https://earthquake.usgs.gov/fdsnws/event/1/query"   

   min_mag = 2                      # the minimum magnitude considered (min_mag)
   start_time = f"{year}-01-01"     
   finish_time = f"{year+1}-01-01"  


   # dictionary of paramenters
   param_dictionary = { "format":"geojson",         
                        "starttime":start_time,
                        "endtime":finish_time,
                        "minmagnitude":min_mag,
                        "latitude": lat,
                        "longitude": lon,
                        "maxradiuskm": 300,
                       }


   maglist = []

   time.sleep(2)                                          # allows time between requests to not burden server 
   result = requests.get(url, params=param_dictionary)    # modifies endpoint according to parameters
   d = result.json()

   # I originally struggled with accessing the json data i wanted, because I was treating the dictionary like a list
   # here was my original code

   # for i in range(0, int(d['metadata']['count'])):
   #       name_of_quake = str(d[i]['features']['properties']['place'])
   #       magnitude = float(d[i]['features']['properties']['mag'])

   # it kept showing keyerror: 0, and I couldn't figure out why
   # in hindsight a dictionary isn't ordered like a list, and there is no "0" key....

   # chat helped me with the first three lines of the loop, and extracted the list within the dictionary for me to work with
   for q in d['features']:
      name_of_quake = q['properties']['place']           # extracts location of quake
      magnitude = q['properties']['mag']                 # extracts magnitude of quake
      maglist.append(magnitude)
      if magnitude == max(maglist):                      # finds largest magnitude quake that year, stores magnitude & place
         maxmag = magnitude
         maxname = name_of_quake
   
   return maxmag, maxname

print(f"What were the largest earthquakes in a 300km radius of Portland, OR every year since 2015?")
print(" ")

lat = 45.523064
lon = -122.676483

for year in range(2015,2025):
   maxmag, maxname = Quake_loop(year, lat, lon)
   print(f"The largest quake in {year} was {maxname}, with a magnitude of {maxmag}")
   print(" ")

What were the largest earthquakes in a 300km radius of Portland, OR every year since 2015?
 
The largest quake in 2015 was 15 km ENE of Cle Elum, Washington, with a magnitude of 4.2
 
The largest quake in 2016 was 4 km W of Hammond, Oregon, with a magnitude of 3.55
 
The largest quake in 2017 was 11 km NE of Hoodsport, Washington, with a magnitude of 4.09
 
The largest quake in 2018 was 37 km NE of CS35_Participant_10 CS35_Participant_9, Washington, with a magnitude of 4.07
 
The largest quake in 2019 was 2 km S of Roosevelt, Washington, with a magnitude of 4.58
 
The largest quake in 2020 was 6 km NNW of Fort Hill Census Designated Place, Oregon, with a magnitude of 3.53
 
The largest quake in 2021 was 6 km NE of Government Camp, Oregon, with a magnitude of 3.94
 
The largest quake in 2022 was 16 km ESE of Lacomb, Oregon, with a magnitude of 4.39
 
The largest quake in 2023 was Port Townsend, WA, with a magnitude of 4.3
 
The largest quake in 2024 was 7 km SSW of River Road, Washingto

#### Largest Portland-centered quakes over the last decade
  + <u>Overall</u>: The API calls to the USGS showed that within 300km of Portland (OR), in the last 10 years
    + the largest quake was of magnitude 4.58, 2km South of Roosevelt, Washington in 2019
    + the smallest quake was of magnitude 3.48, 7km South/Southwest of River Road, Washington in 2024 <br><br>
  + <u>Reflection</u>: It seems that not many large earthquakes have been happening in the area for a while (though, not sure 10 years is "a while" in geological time). I have always grown up hearing that a "big one" was going to hit the PNW though. Not sure if this data means anything towards that. <br><br>
  + <u>Opportunities</u>: Using this code for different locations, making the time range longer, looking for smallest earthquakes <br><br>

In [41]:
# compares the largest quakes from the last ten years in two locations
import requests
import time

def Quake_compare(LA, PDX):
   countLA = 0
   countPDX = 0
   
   for year in range(2015,2025):                                 # each year, over last ten years
      maxmagLA, maxnameLA = Quake_loop(year, LA[0], LA[1])       # find largest LA quake
      maxmagPDX, maxnamePDX = Quake_loop(year, PDX[0], PDX[1])   # and largest Portland quake

      if maxmagLA >= maxmagPDX:       # if largest quake was in LA, keep count of that
         countLA +=1
      else: 
         countPDX +=1                 # also keep count if largest quake was in Portland
   return countLA, countPDX

LA = [34.052235,-118.243683]
PDX = [45.523064, -122.676483]

countLA, countPDX = Quake_compare(LA, PDX)

if countLA > countPDX:      # now compare which had the larger quake over all ten years
   print(f"There were more large earthquakes in LA than Portland")
else:  print(f"There were more large earthquakes in Portland than LA")

There were more large earthquakes in LA than Portland


#### Portland-centered vs LA-centered quakes over the last decade
  + <u>Overall</u>: The API calls to the USGS compared quakes within 300km of Portland (OR) and LA in the last 10 years
    + LA overall had larger earthquakes per year than Portland <br><br>
  + <u>Reflection</u>: This makes sense, given what I've experienced living around LA and Portland<br><br>
  + <u>Opportunities</u>: consider different locations, have the code read out the larger of the largest earthquakes per year  <br><br>

#### Final API challenge:  an open-ended, two-hop API task...

<font color="Coral">Key constraint: _Use two hops!_</font>  &nbsp;&nbsp; Be sure to dome something that uses two separate API calls, in which the results of the first affect the second - culminating in an aggregate, overall result.

Possibilities: 
  + You should use at least one other API. (See the hw2 page for links to many.)
      + The [Poke API](https://pokeapi.co/)  or [one of these](https://medium.com/codex/15-fun-and-interesting-apis-to-use-for-your-next-coding-project-in-2022-86a4ff3a2742) or [one of the many, many more!](https://github.com/public-apis/public-apis)
  + That said, <u><b>one</b></u> of the two can be ISS or USGS, as you used above.



<font size="-2">PS. My example of this is in ``superbowl_prediction.ipynb`` with the starter files. It used to be here, but it was too clutter-y... .</font>

<br><hr><br>

Big-picture, an important part of using API calls is gathering a _specific piece_ of data from an otherwise too-large ocean of raw material. That element then allows you to express a natural _next_ specific question, obtain its relevant data, and continue to build from there, with each round-trip branching into additional (possible) questions... .

<br>

Good luck API'ing!!

<br>



In [ ]:
# preevolution chain 
import requests

def Evolves_from(pokemon):
    url= f'https://pokeapi.co/api/v2/pokemon-species/{pokemon}'       # modifies endpoint according to pokemon of choice
    result = requests.get(url)
    pokemond = result.json()
    #print(pokemon)

    if pokemond["evolves_from_species"] == None:                      # sees if your pokemon has a preevolution or not
        prepokemon = None
    else: prepokemon = pokemond["evolves_from_species"]['name']

    return prepokemon

pokemon = input("What pokemon do you want to find the preevolution(s) of? ")

while Evolves_from(pokemon) != None:                             # loops the function until there are no more preevolutions & you reached base evolution
    print(f"{pokemon} evolves from {Evolves_from(pokemon)}")
    pokemon = Evolves_from(pokemon)

if Evolves_from(pokemon) == None:
    print(f"{pokemon} does not have a pre-evolution")

raichu evolves from pikachu
pikachu evolves from pichu
pichu does not have a pre-evolution


In [ ]:
# getting "flavor text" (small blurbs about a pokemon) for pokemon & its preevolution chain

def flavor_text(pokemon):
    url= f'https://pokeapi.co/api/v2/pokemon-species/{pokemon}'        # modifies endpoint according to pokemon of choice
    result = requests.get(url)
    pokemond = result.json()
    
    flavortext = pokemond["flavor_text_entries"][0]['flavor_text']     # extracts flavor text

    return flavortext

pokemon = input("What pokemon, and its preevolution(s), do you want to know more about? ")

flavortext = flavor_text(pokemon)       # prints flavor text of originally selected pokemon
print(f"{pokemon}: {flavortext}")
print(" ")

while Evolves_from(pokemon) != None:                    # loops while your pokemon has preevolutions          
    flavortext = flavor_text(Evolves_from(pokemon))     # to find their flavor texts as well

    print(f"{Evolves_from(pokemon)}: {flavortext}")
    print(" ")
    pokemon= Evolves_from(pokemon)

if Evolves_from(pokemon) == None:                       # at the end of the chain, prints that there are no more preevolutions
    print(f"{pokemon} does not have a pre-evolution")


gengar: Under a full moon,
this POKéMON
likes to mimicthe shadows of
people and laugh
at their fright.
 
haunter: Because of its
ability to slip
through blockwalls, it is said
to be from an­
other dimension.
 
gastly: Almost invisible,
this gaseous
POKéMON cloaksthe target and
puts it to sleep
without notice.
 
gastly does not have a pre-evolution
